Add more FAQs (#1990)
emqtt/emq-docs-cn: en_US/faq/use-guide.md

- `proto_unexpected_c`：Repeatedly received an MQTT connection request when there is already an MQTT connection

- `idle_timeout`： After the TCP connection is established for 15s, the connect packet has not been received yet.

> ## Can EMQX guarantee the original order when forwarding the messages to subscribers?

Reviewer: ```suggestion
## Can EMQX guarantee the original order when forwarding messages to subscribers?
```

**Tags:** [*Pub/Sub*](tags.md#pub-sub)

> EMQX will ensure that messages with the same topic from the same client are forwarded in the order of their arrival, regardless of the QoS level of the messages. The QoS level does not affect the order of message forwarding. Whether messages are lost or duplicated, they will not cause message disorder. This is also the requirement from MQTT.

Reviewer: ```suggestion
EMQX ensures that messages with the same topic from the same client are forwarded in the order they were received, regardless of the QoS level. The message forwarding order remains consistent regardless of message loss or duplication, as per MQTT requirements.
```

> However, EMQX does not guarantee the forwarding order of messages from different topics. We can consider them as entering different channels. For example, messages from topic A arrive at EMQX before messages from topic B, it is possible that messages from topic B will be forwarded earlier.

Reviewer: ```suggestion
However, EMQX does not guarantee the forwarding order of messages from different topics. These messages can be considered as entering separate channels. For example, if messages from topic A arrive at EMQX before messages from topic B, it is possible that messages from topic B will be forwarded earlier.
```

## What should I do when I encounter problems related to client connection, publishing and subscribing?

**Tags:** [*Debug*](tags.md#debug)

> EMQX's debug logs already record all the behaviors and phenomena. By reading the debug logs, we can know when the client initiated the connection, which parameters were specified during the connection, whether the connection was successful, and the reasons for connection rejection, etc. However, due to the excessive information recorded in debug logs, it may incur additional resource consumption and make it difficult to analyze individual clients or topics.

Reviewer: ```suggestion
EMQX's debug logs already capture all the behaviors and phenomena. By viewing the debug logs, we can determine when the client initiated the connection, the parameters specified during the connection, the success of rejection of the connection, and the reasons for rejection, among other details. However, the extensive information logged in debug mode can consume additional resources and make it challenging to analyze individual clients or topics.
```

> So EMQX provides a [Log Trace](../getting-started/log.md) feature. We can specify the clients or topics we want to trace, and EMQX will output all the debug logs related to those clients or topics to the designated log file. This makes it convenient for both self-analysis and seeking help from the community.

Reviewer: ```suggestion
So EMQX provides a [Log Trace](../getting-started/log.md) feature. We can specify the clients or topics we want to trace, and EMQX will output all the debug logs related to those clients or topics to the designated log file. This makes it convenient for both self-analysis and seeking help from the community.
```
```suggestion
So EMQX provides a [Log Trace](../getting-started/log.md) feature. We can specify the clients or topics we want to trace, and EMQX will output all the debug logs related to those clients or topics to the designated log file. This makes it convenient for both self-analysis and seeking help from the community.
```
```suggestion
To address this, EMQX provides a [Log Trace](../getting-started/log.md) feature. We can specify the clients or topics we want to trace, and EMQX will output all the debug logs related to those clients or topics to the designated log file. This facilitates self-analysis and seeking assistance from the community.
```

> Just so you know, if the client cannot connect to EMQX for network reasons, the log tracing feature will not be helpful, because EMQX has not received any messages in such cases. This situation often occurs due to network configuration issues such as firewalls or security groups, which result in the server port not being open. This is particularly common when deploying EMQX on cloud instances. So, in addition to log tracing, we can troubleshoot network-related issues by checking port occupation, listening status, and network configurations.

Reviewer: ```suggestion
It's important to note that if the client cannot establish a connection with EMQX due to network issues, the log tracing feature will not be useful since EMQX does not receive any messages in such cases. This situation often arises from network configuration problems like firewalls or security groups, resulting in closed server ports. This is particularly common when deploying EMQX on cloud instances. Therefore, in addition to log tracing, troubleshooting network-related issues involves checking port occupation, listening status, and network configurations.
```

## Why are there client IDs like "CENSYS" or other unfamiliar clients?

**Tags:** [*Security*](tags.md#security)

> CENSYS is an internet scanning and reconnaissance tool that periodically scans the IPv4 address space to discover default ports for protocols such as HTTP, SSH, MQTT, etc. So if you find MQTT clients with a client ID of "CENSYS" or other unknown clients accessing your MQTT broker, it indicates that you are currently at a relatively lower level of security protection. The following measures can effectively help you avoid this issue:

Reviewer: ```suggestion
CENSYS is an internet scanning and reconnaissance tool that performs regular scans of the IPv4 address space to identify default ports for various protocols such as HTTP, SSH, MQTT, and etc. Therefore, if you notice MQTT clients with a client ID of "CENSYS" or other unfamiliar clients accessing your MQTT broker, it indicates a relatively lower level of security protection. To address this issue effectively, consider implementing the following measures:
```

1. Avoid using default configurations, such as the AppID and AppSecret used for verifying HTTP API access permissions in EMQX.

> 2. Enable authentication, which can be password-based authentication or JWT authentication, to prevent situations where only knowledge of an IP address is required for login.

Reviewer: ```suggestion
2. Enable authentication mechanisms like password-based authentication or JWT authentication to prevent unauthorized access where only knowledge of an IP address is sufficient for login.
```

> 3. Enable TLS mutual authentication, allowing only clients with valid certificates to access the system.

Reviewer: ```suggestion
3. Implement TLS mutual authentication to allow access only to clients with valid certificates.
```

> 4. Enable authorization to prevent unauthorized devices from accessing sensitive data after logging in.

Reviewer: ```suggestion
4. Enable proper authorization mechanisms to restrict access to sensitive data for unauthorized devices.
```

5. Configure your firewall to close unnecessary ports as much as possible.

> ## Can I shared subscribe to the system topic?

Reviewer: ```suggestion
## Can I subscribe to system topics using shared subscriptions?
```

**Tags:** [*Shared Subscription*](tags.md#shared-subscription)

> Yes, some system messages, such as client online/offline events, may be published frequently, so using a shared subscription is essential for clients. Example subscription: `$share/group1/$SYS/brokers/+/clients/+/connected`.

Reviewer: ```suggestion
Yes, it is possible to use shared subscriptions to subscribe to certain system messages, such as client online/offline events, which are published frequently. Shared subscriptions are particularly useful for clients in such cases. For instance, you can subscribe to the following topic using a shared subscription: $share/group1/$SYS/brokers/+/clients/+/connected.
```

## Why can't I receive retained messages when using shared subscriptions?

**Tags:** [*Shared Subscription*](tags.md#shared-subscription)

> The MQTT protocol stipulates that when the client uses a shared subscription, the server cannot send retained messages to it.

Reviewer: ```suggestion
According to the MQTT protocol, when a client uses a shared subscription, the server is not allowed to send retained messages to that client.
```

> ## Why are messages occasionally lost when using shared subscriptions?

Reviewer: ```suggestion
## Why do messages sometimes get lost when using shared subscriptions?
```

**Tags:** [*Shared Subscription*](tags.md#shared-subscription)

> When the connection of the shared subscriber is disconnected but the session still exists, the server will continue delivering the messages to the shared subscriber, and these messages will be temporarily stored in the corresponding session. This will cause other online shared subscribers to appear to have not consumed all the messages. In addition, if the shared subscriber chooses to create a new session when reconnecting, the messages cached in the old session will be permanently lost.

Reviewer: ```suggestion
When a shared subscriber's connection is disconnected but the session remains active, the server continues to deliver messages to the subscriber, which are temporarily stored in the session. As a result, other active shared subscribers may appear as if they have not consumed all the messages. In addition, if the shared subscriber chooses to create a new session when reconnecting, the messages cached in the old session will be permanently lost.
```

> If we confirm that the above situation does not exist, but the problem of message loss still exists, then we can use the client tracking function of EMQX for further investigation.

Reviewer: ```suggestion
If you have verified that the aforementioned situation does not occur, yet the issue of message loss persists, you can use the client tracking feature of EMQX to conduct further investigation.
```

> ## What should I do if EMQX prompts that the port is occupied (eaddrinuse) when starting?

Reviewer: ```suggestion
## What should I do if EMQX reports a "port is occupied (eaddrinuse)" error during startup?
```

Author: “port is occupied” is not an actual log.

**Tags:** [*Failed to Start*](tags.md#failed-to-start)

By default, EMQX will occupy 7 ports when it starts. They are:

> 1. 1883, used for MQTT over TCP listener, can be modified by configuration

Reviewer: ```suggestion
1. Port 1883, used for MQTT over TCP listener. It can be modified through configuration.
```

> 2. 8883, used for MQTT over SSL/TLS listener, can be modified by configuration

Reviewer: ```suggestion
2. Port 8883, used for MQTT over SSL/TLS listener. It can be modified through configuration.
```

> 3. 8083, used for MQTT over WebSocket listener, can be modified by configuration

Reviewer: ```suggestion
3. Port 8083, used for MQTT over WebSocket listener. It can be modified through configuration.
```

> 4. 8084, for MQTT over WSS (WebSocket over SSL) listener, can be modified by configuration

Reviewer: ```suggestion
4. Port 8084, used for MQTT over WSS (WebSocket over SSL) listener. It can be modified through configuration.
```

> 5. 18083, the default listening port of the HTTP API service. And the dashboard also depends on this port, which can be modified by configuration

Reviewer: ```suggestion
5. Port 18083, the default listening port for the HTTP API service. The dashboard also relies on this port, which can be modified through configuration.
```

> 6. 4370, used for remote function calls of EMQX distributed cluster, Mnesia data synchronization, etc. This port will be occupied by default even if no cluster is formed. The listening port should be `BasePort (4370) + Offset`, 4370 is fixed and cannot be modified, and Offset is determined by the numeric suffix of the Name part in the node name (`Name@Host`). If there is no numeric suffix, it defaults to 0. For example, the Offset of `emqx@127.0.0.1` is 0, and the Offset of `emqx1@127.0.0.1` is 1.

Reviewer: ```suggestion
6. Port 4370, used for remote function calls in the EMQX distributed cluster and Mnesia data synchronization. This port is occupied by default, even if no cluster is formed. The listening port is determined by `BasePort (4370) + Offset`, where 4370 is fixed and cannot be modified, and Offset is determined by the numeric suffix in the node name (`Name@Host`). If there is no numeric suffix, it defaults to 0. For example, the Offset for `emqx@127.0.0.1` is 0, and the Offset for `emqx1@127.0.0.1` is 1.
```

> 7. 5370, the cluster RPC port used to share the pressure of the last port, mainly used for forwarding MQTT messages between nodes. Similar to port 4370, even if no cluster is formed, this port will be occupied by default, and it should actually be `BasePort (5370) + Offset`, 5370 is fixed and cannot be modified, and Offset is determined by the Name part of the node name (`Name@Host`). If there is no number suffix, it defaults to 0.

Reviewer: ```suggestion
7. Port 5370, the cluster RPC port used for load sharing. It is mainly used for forwarding MQTT messages between nodes. Similar to port 4370, this port is occupied by default, even if no cluster is formed. The actual listening port is `BasePort (5370) + Offset`, where 5370 is fixed and cannot be modified, and Offset is determined by the Name part of the node name (`Name@Host`). If there is no numeric suffix, it defaults to 0.
```

> ## Why does EMQX output the log "WARNING: Default (insecure) Erlang cookie is in use." when it starts?

Reviewer: ```suggestion
## Why does EMQX output the log "WARNING: Default (insecure) Erlang cookie is in use." during startup?
```

**Tags:** [*Log*](tags.md#log)

The complete WARNING log is as follows:

```
WARNING: Default (insecure) Erlang cookie is in use.
WARNING: Configure node.cookie in /usr/lib/emqx/etc/emqx.conf or override from environment variable EMQX_NODE__COOKIE
WARNING: NOTE: Use the same cookie for all nodes in the cluster.
```

> Only EMQX nodes using the same cookie can form a cluster. While a cookie does not secure cluster communication, it prevents a node from connecting to a cluster it did not intend to communicate with. EMQX nodes uniformly use `emqxsecretcookie` as cookie by default, so we recommend users change the cookie value when building a cluster.

Reviewer: ```suggestion
Only EMQX nodes using the same cookie can form a cluster. While a cookie does not secure cluster communication, it prevents a node from connecting to a cluster it did not intend to communicate with. By default, EMQX nodes uniformly use the cookie value `emqxsecretcookie`. However, we recommend that users change the cookie value when building a cluster to enhance security.
```

> The second warning log indicates two ways to modify the cookie: `node.cookie` in the `emqx.conf` configuration file and the environment variable `EMQX_NODE__COOKIE`.

Reviewer: ```suggestion
The second warning log indicates two ways to modify the cookie: by editting `node.cookie` in the `emqx.conf` configuration file or by setting the environment variable `EMQX_NODE__COOKIE`.
```

> ## When using Docker to deploy EMQX, why does the restart of the container cause data loss, such as configured rules and resources?

Reviewer: ```suggestion
## Why does restarting the EMQX Docker container cause data loss, such as configured rules and resources?
```

**Tags:** [*Persistence*](tags.md#persistence)

> The runtime data of EMQX are all stored in the `/opt/emqx/data` directory, such as the configuration of rules and resources, retained messages, etc. We need to mount the `/opt/emqx/data` directory to a local host directory or a data volume to ensure that these data will not be lost due to container restarts.

Reviewer: ```suggestion
The runtime data of EMQX is stored in the `/opt/emqx/data` directory, including configuration rules, resources, retained messages, etc. To ensure data persistence during container restarts, it's important to mount the `/opt/emqx/data` directory to a local host directory or a data volume.
```

> But we may find that even if the `/opt/emqx/data` directory has been mounted, the data may still be lost after the container restarts. This is because the runtime data of EMQX is stored in the `/opt/emqx/data/mnesia/${Node Name}` directory. So the data loss is actually that the node name of EMQX has changed after the container is restarted, which in turn caused EMQX to create a new storage directory.

Reviewer: ```suggestion
However, even if the `/opt/emqx/data` directory is properly mounted, data loss may still occur after container restarts. This is because the runtime data of EMQX is stored in the `/opt/emqx/data/mnesia/${Node Name}` directory, and when the container is restarted, the node name of EMQX changes, leading to the creation of a new storage directory.
```

> EMQX node name consists of Name and Host, where Host comes from the container's IP address by default. Under the default network configuration, the container's IP may change after restarting, so we have to keep the container's IP fixed.

Reviewer: ```suggestion
EMQX node name consists of Name and Host, with the Host derived from the container's IP address by default. Under the default network configurations, the container's IP may change upon restarting, so you need to maintain a fixed IP for the container.
```

> EMQX provides an environment variable `EMQX_HOST`, which allows us to set the Host part of the node name. Of course, the premise is that this Host must be connectable to other nodes, so we also need to use it with the network alias:

Reviewer: ```suggestion
To address this issue, EMQX provides an environment variable, `EMQX_HOST`, which allows you to set the Host part of the node name. However, it is crucial that this Host value is reachable by other nodes, so it should be used in conjunction with a network alias. Here is an example command for running the EMQX Docker container with the EMQX_HOST environment variable and a network alias:
```

```
docker run -d --name emqx -p 18083:18083 -p 1883:1883 -e EMQX_HOST=alias-for-emqx --network example --network-alias alias-for-emqx --mount type=bind,source=/tmp/emqx,target=/opt/emqx/data emqx:5.0.24
```

> ## Reasons for common logs

Reviewer: ```suggestion
## Common Log Reasons in EMQX
```

**Tags:** [*Log*](tags.md#log)

### Keyword: “reason: {shutdown, takeovered}”

> In EMQX, each MQTT client connection is maintained by an independent process. Whenever an MQTT client connects, EMQX will create a new process. If the client wants to restore communication from the existing session state, EMQX needs to migrate the session state from the old process to the new one. We call this process: Take over, and the old process will be closed after the takeover is completed, and the reason for closing is `{shutdown, takeovered}`. If the original client is still connected, EMQX will send a DISCONNECT packet with Reason Code 0x8E (Session taken over), and then close the old connection.

Reviewer: ```suggestion
In EMQX, each MQTT client connection is maintained by an independent process. When a client connects, EMQX creates a new process. If the client wants to resume communication from an existing session state, EMQX needs to migrate the session state from the old process to the new one. This process is called "takeover," and the old process is closed after the takeover is completed with the reason `{shutdown, takeovered}`. If the original client is still connected, EMQX sends a DISCONNECT packet with Reason Code 0x8E (Session taken over) before closing the old connection.
```

### Keyword: “reason: {shutdown, discarded}”

> Contrary to session takeover, if the client indicates that it wants to start a new session when connecting, then EMQX needs to discard the old session state, and EMQX will close the old process with the reason `{shutdown, discarded}`. If the original client is still connected, EMQX will send a DISCONNECT packet with Reason Code 0x8E (Session taken over), and then close the old connection.

Reviewer: ```suggestion
In contrast to session takeover, if a client indicates that it wants to start a new session upon connecting, EMQX needs to discard the old session state. EMQX closes the old process with the reason `{shutdown, discarded}`. If the original client is still connected, EMQX sends a DISCONNECT packet with Reason Code 0x8E (Session taken over) before closing the old connection.
```

### Keyword: “reason: {shutdown, kicked}”

> Indicates that the client connection was manually kicked out (click the Kick Out button on the Dashboard or call the `DELETE clients/{clientid}` API). If the original client is still connected, EMQX will send a DISCONNECT packet with Reason Code 0x98 (Administrative action), and then close the old connection.

Reviewer: ```suggestion
Indicates that the client connection was manually kicked out, either by clicking the "Kick Out" button on the Dashboard or by calling the `DELETE clients/{clientid} API`. If the original client is still connected, EMQX sends a DISCONNECT packet with Reason Code 0x98 (Administrative action) before closing the old connection.
```

### Keyword: “reason: {shutdown, tcp_closed}”

This log indicates that the client directly closed the network connection without sending a DISCONNECT packet. The corresponding reason would be `{shutdown, normal}`, which means that the client first sent a DISCONNECT packet with a Reason Code of 0 and then closed the network connection.

> Therefore, when you encounter that `tcp_closed` appears in the EMQX logs, and this disconnection behavior is not what you expected, it is recommended to investigate the client's implementation for any potential issues.

Reviewer: ```suggestion
If you encounter the `tcp_closed` keyword in the EMQX logs and this disconnection behavior is not expected, it is recommended to investigate the client's implementation for any potential issues.
```

### Keyword: “maximum heap size reached”

> The appearance of this log indicates that the heap occupied by the current client process has exceeded the preset maximum value, and the client process will be forcibly killed to ensure the availability of EMQX, and avoid the unlimited growth of the memory usage of the client process and eventually lead to EMQX OOM. 

Reviewer: ```suggestion
This log appears when the heap occupied by the current client process exceeds the preset maximum value. The client process will be forcibly terminated to ensure EMQX's availability and to prevent excessive memory usage leading to EMQX being Out of Memory (OOM). 
```

> Usually, message accumulation is the main reason for the increase in client process heap occupation, and message accumulation is generally because the client's consumption capacity does not match the peer's message production capacity. The recommended solution is to optimize the client processing code or use shared subscriptions to disperse the load.

Reviewer: ```suggestion
Usually, message accumulation is the main cause of increased heap occupation, often due to the client's consumption capacity not matching the peer's message production capacity. It is recommended to optimize the client processing code or use shared subscriptions to distribute the load.
```

> The related configuration item is `force_shutdown_policy`, and its format is `<Maximum Message Queue Length>|<Maximum Heap Size>`, for example, `10000|64MB`.  `<Maximum Heap Size>` refers to the maximum heap memory that each client process can occupy.

Reviewer: ```suggestion
The related configuration item is `force_shutdown_policy`, with the format `<Maximum Message Queue Length>|<Maximum Heap Size>`, for example, `10000|64MB`. `<Maximum Heap Size>` refers to the maximum heap memory that each client process can occupy.
```

### Keyword: “Parse failed for function_clause”

> This log indicates that the parsing of the packet failed. Maybe because it is not an MQTT packet, we have encountered many situations where HTTP requests are sent to the MQTT port, or because the packet contains non-UTF-8 characters. We can search the `Frame data` keyword in this "Parse failed..." log to view the complete message and help us analyze the possible reasons for the parsing failure.

Reviewer: ```suggestion
This log indicates that the parsing of the packet failed. Maybe because it is not an MQTT packet, we have encountered many situations where HTTP requests are sent to the MQTT port, or because the packet contains non-UTF-8 characters. We can search the `Frame data` keyword in this "Parse failed..." log to view the complete message and help us analyze the possible reasons for the parsing failure.
```
```suggestion
This log indicates a failure in packet parsing. It may occur when a non-MQTT packet is sent to the MQTT port, or when the packet contains non-UTF-8 characters. Searching for the `Frame data` keyword in the `Parse failed...` log can provide the complete message and help analyze the possible reasons for the parsing failure.
```

### Keyword: “Dropped msg due to mqueue is full”

> EMQX can send multiple unconfirmed QoS 1 and QoS 2 messages at the same time, but limited by the performance of the client, we usually limit the maximum number of messages allowed to be sent at the same time. After reaching this limit, subsequent arriving messages will be cached by EMQX in the message queue of each client process.

Reviewer: ```suggestion
EMQX can send multiple unconfirmed QoS 1 and QoS 2 messages simultaneously. However, to prevent excessive message buffering, there is a maximum length limit for the message queue. When reaching this limit, subsequent arriving messages will be cached by EMQX in the message queue of each client process.
```

Author: The original meaning is changed here. I made the following changes:

```
EMQX can send multiple unconfirmed QoS 1 and QoS 2 messages simultaneously, but limited by the performance of the client, we usually limit the maximum number of messages in flight. After reaching this limit, subsequent arriving messages will be cached by EMQX in the message queue of each client process.

However, the message queus still has a length limit. So If the message queue is full when a new message arrives, the latest message will be enqueued, but the oldest message in the queue will be dequeued and dropped. EMQX generates the log message: "Dropped msg due to mqueue is full."
```

Reviewer: OK, just made a little editing:

"EMQX can send multiple unconfirmed QoS 1 and QoS 2 messages simultaneously, but limited by the performance of the client, we usually limit the maximum number of messages in flight. After reaching this limit, subsequent incoming messages will be cached by EMQX in the message queue of each client process.

However, the message queues still have a length limit. If the message queue is full when a new message arrives, the most recent message will be enqueued, but the oldest message in the queue will be dequeued and discarded. EMQX generates the log message: "Dropped msg due to mqueue is full."

Author: Updated, I still use 'dropped' rather than 'discarded', keep it consistent with the log.

> To prevent excessive message buffering, EMQX also sets a maximum length limit for the message queue. If the message queue is full when a message arrives, the latest message will still be enqueued, but the oldest message in the queue will be dequeued and dropped, and EMQX will generate the log message: "Dropped msg due to mqueue is full."

Reviewer: ```suggestion
If the message queue is full when a new message arrives, the latest message will be enqueued, but the oldest message in the queue will be dequeued and dropped. EMQX generates the log message: "Dropped msg due to mqueue is full."
```

> If this log only appears during traffic peaks, you can modify the configuration item `max_mqueue_len` only to increase the maximum length of the message queue. However, if it continues to appear, it indicates that the current client consumption capability is poor. In such cases, it is recommended to optimize the client's code or use shared subscriptions to distribute the load.

Reviewer: ```suggestion
If this log only appears during traffic peaks, you can increase the maximum length of the message queue by modifying the `max_mqueue_len` configuration item. However, if it continues to appear, it indicates poor client consumption capability. In such cases, it is recommended to optimize the client's code or use shared subscriptions to distribute the load.
```
